src/viewport_renderer.rs

```diff
@@ -0,0 +1,402 @@
+        self.aspect_ratio = width as f32 / height as f32;
+
+        let matrix = Self::create_matrix(self.aspect_ratio, &self.view_matrix);
+        Self::upload_uniform_matrix_buffer(device, &self.uniform_matrix_buffer, &matrix);
+
+        let depth_texture = Self::create_depth_texture(device, tex_width, tex_height);
+        self.depth_texture = depth_texture.create_default_view();
+    }
+
+    pub fn add_geometry(
+        &mut self,
+        device: &wgpu::Device,
+        vertices: &[Vertex],
+    ) -> Result<GeometryId, ViewportRendererAddGeometryError> {
+        let id = GeometryId(self.geometries_next_id);
+        let vertex_count = u32::try_from(vertices.len()).map_err(|_| {
+            ViewportRendererAddGeometryError::TooManyVertices(u32::max_value(), vertices.len())
+        })?;
+
+        log::debug!("Adding geometry {} with {} vertices", id.0, vertex_count);
+
+        let vertex_buffer = device
+            .create_buffer_mapped(vertices.len(), wgpu::BufferUsage::VERTEX)
+            .fill_from_slice(vertices);
+
+        self.geometries.insert(
+            id.0,
+            GeometryDescriptor {
+                vertices: (vertex_buffer, vertex_count),
+                indices: None,
+            },
+        );
+
+        self.geometries_next_id += 1;
+        Ok(id)
+    }
+
+    pub fn add_geometry_indexed(
+        &mut self,
+        device: &wgpu::Device,
+        vertices: &[Vertex],
+        indices: &[Index],
+    ) -> Result<GeometryId, ViewportRendererAddGeometryError> {
+        let id = GeometryId(self.geometries_next_id);
+        let vertex_count = u32::try_from(vertices.len()).map_err(|_| {
+            ViewportRendererAddGeometryError::TooManyVertices(u32::max_value(), vertices.len())
+        })?;
+        let index_count = u32::try_from(indices.len()).map_err(|_| {
+            ViewportRendererAddGeometryError::TooManyIndices(u32::max_value(), indices.len())
+        })?;
+
+        log::debug!(
+            "Adding indexed geometry {} with {} vertices and {} indices",
+            id.0,
+            vertex_count,
+            index_count,
+        );
+
+        let vertex_buffer = device
+            .create_buffer_mapped(vertices.len(), wgpu::BufferUsage::VERTEX)
+            .fill_from_slice(vertices);
+
+        let index_buffer = device
+            .create_buffer_mapped(indices.len(), wgpu::BufferUsage::INDEX)
+            .fill_from_slice(indices);
+
+        self.geometries.insert(
+            id.0,
+            GeometryDescriptor {
+                vertices: (vertex_buffer, vertex_count),
+                indices: Some((index_buffer, index_count)),
+            },
+        );
+
+        self.geometries_next_id += 1;
+        Ok(id)
+    }
+
+    pub fn remove_geometry(&mut self, id: GeometryId) {
+        log::debug!("Removing geometry with {}", id.0);
+        // Dropping the geometry descriptor here unstreams the buffers from device memory
+        self.geometries.remove(&id.0);
```

Comment:
Wouldn't it be a good idea to also implement `Drop` which would remove all geometry?

Reply:
I think that happens automatically when the `HashMap` is dropped.